src/java.desktop/share/classes/javax/swing/text/html/CSS.java

```diff
@@ -2604,9 +2672,21 @@ Object toStyleConstants(StyleConstants key, View v) {
             return Float.valueOf(getValue(false));
         }
 
+        @Override
+        public int hashCode() {
+            return Float.hashCode(span);
+        }
+
+        @Override
+        public boolean equals(Object val) {
+            return val instanceof CSS.LengthValue lu
+                   && Objects.equals(svalue, lu.svalue);
+        }
```

Comment:
Why is this case still using the string and not 'span' ?

Reply:
This is because it can cater to both units like px,pt and %
For example, for attributes
`{"margin-top: 42px", "margin-top: 22px"}`

`span `22.0, 42.0
`percentage `false
`units `px

where we can use `span == lu.span && units == lu.units` check
but for attributes like
`{"margin-top: 100%", "margin-top: 200%"}`

`span `1.0
`percentage `true
`units `null (where we need to use the string so for optimization I have used the same check for both)
I guess span is 1.0 is because of this code 
https://github.com/openjdk/jdk/blob/325940b0919e649e84358db35a55b3ac5dfe0acf/src/java.desktop/share/classes/javax/swing/text/html/CSS.java#L2557-L2561

Reply:
Removed optimization for non-percentage units which I guess will be more used and it will also do away with `{"margin-top: 42px", "margin-top: 42 px"}` space issue